### PR TITLE
[SYCL][Driver] Fix -fgpu-rdc option for CUDA

### PR DESCRIPTION
**Pull request by GeorgeWeb**
This commit fixes an issue with -fgpu-rdc for NVPTX targets in the SYCL driver. When the driver calls the `ptxas` assembling job in relocatable device code compilation mode, we pass `-c` to the command-line to get a relocatable code ELF instead of a statically fully linked object. This is the correct behaviour for the purpose of CUDA separate compilation flow with RDC, but in SYCL it does not make sense for us to do so because we do RDC at a bitcode level rather than PTX. We link the IR modules with llvm-link way earlier and the linking method based on `-fgpu[no]-rdc` is determined there. Hence, we always want a fully linked object from `ptxas` in SYCL. This simplifies the need for adding a `nvlink` job in the pipeline.

Related github issue: https://github.com/intel/llvm/issues/8690

### CODE REVIEW

**Comment by GeorgeWeb** [comment]
**For further discussion:** Maybe in the future we should consider adding a `nvlink` job which is something I started here: https://github.com/intel/llvm/pull/11400, and think about a different compilation where we postpone linking for NVPTX/CUDA modules to later point for device linking which would mean supporting a new compilation path for NVPTX targets (e.g. no LTO / IR linking and nvlink the TU objects later). Another reason for an `nvlink` stage can be adding the ability for the user to link external static libraries/archives that contain cubin device codes which I don't think we support.

**Comment by GeorgeWeb** [comment]
> Do you know if there could be a similar issue in the AMD driver when using SYCL offload?

As far as I understand it, the AMD-HIP driver invokes `lld` and links the bitcode. i.e.:
```bash
"/llvm/build/bin/llvm-foreach" "--out-ext=out" "--in-file-list=/tmp/sycl_fgpurdc-gfx1030-afa93a.o" "--in-replace=/tmp/sycl_fgpurdc-gfx1030-afa93a.o" "--out-file-list=/tmp/sycl_fgpurdc-gfx1030-1930f1.out" "--out-replace=/tmp/sycl_fgpurdc-gfx1030-1930f1.out" "--" "/home/georgim/workspaces/forks/llvm/build/bin/lld" "-flavor" "gnu" "-m" "elf64_amdgpu" "--no-undefined" "-shared" "-plugin-opt=-amdgpu-internalize-symbols" "-plugin-opt=mcpu=gfx1030" "--whole-archive" "-o" "/tmp/sycl_fgpurdc-gfx1030-1930f1.out" "/tmp/sycl_fgpurdc-gfx1030-afa93a.o" "--no-whole-archive"
```

---

WRT Cuda:
So there can be an argument to also allow `ptxas -c` for Cuda-RDC and `nvlink` the `.cubin`s containing the PTX code, but there is no point as of yet in the way our driver is set up. It's a good discussion and something we may want to handle too later.

File: clang/lib/Driver/ToolChains/Cuda.cpp [hunk 1]
```diff
@@ -476,11 +476,15 @@ void NVPTX::Assembler::ConstructJob(Compilation &C, const JobAction &JA,
...
-           JA.isOffloading(Action::OFK_SYCL))
+  else if (JA.isOffloading(Action::OFK_Cuda))
     // In CUDA we generate relocatable code by default.
     Relocatable = Args.hasFlag(options::OPT_fgpu_rdc, options::OPT_fno_gpu_rdc,
                                /*Default=*/false);
+  else if (JA.isOffloading(Action::OFK_SYCL))
+    // In SYCL we control [no-]rdc linking at bitcode stage with 'llvm-link'.
+    // This allows for link-time optimisations and for now we do no support a
+    // non-LTO path, which means we cannot generate relocatable device code.
+    Relocatable = false;
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/Cuda.cpp:487]: could a test be added to verify the non-addition of `-c` for the SYCL case?

Comment (GeorgeWeb) [reply clang/lib/Driver/ToolChains/Cuda.cpp:487]: Sure! Added. :)

**Comment by sarnex** [comment]
@GeorgeWeb Sorry, just found this PR. So does this mean we don't honor `-fno-sycl-rdc` for CUDA?

**Comment by GeorgeWeb** [comment]
> @GeorgeWeb Sorry, just found this PR. So does this mean we don't honor `-fno-sycl-rdc` for CUDA and the SYCL driver acts like `-fsycl-rdc` in terms of how it does device linking? Or does it mean we just aren't passing `-c` to the backend tool but do everything else the same in the SYCL driver?

Hey @sarnex. Exactly the second 👍. We honour both options for RDC and all linking is done at bitcode/IR level, linking IR modules before getting the fat-binaries for Cuda. So, this change just means we are not passing "-c" to the backend tools of the Cuda/NVPTX toolchain that work on the target device code (PTX and SASS assembly), but the rest is the same in the driver. 


Also, as I tried to expand in one of my comments - in order to do proper Cuda RDC device code linking of the PTX objects, we will have to refactor the driver a little bit, at least for NVPTX, such that we loose LTO and the llvm-link stage that links the separate TUs together but that's a separate concern.

**Review by sarnex — APPROVED** [review]
@GeorgeWeb LGTM, glad it's the second one.

**Comment by againull** [comment]
@GeorgeWeb Could you please address CI failure on Windows?
FAIL: Clang :: Driver/sycl-cuda-rdc.cpp (10320 of 18711)
******************** TEST 'Clang :: Driver/sycl-cuda-rdc.cpp' FAILED ********************
Exit Code: 1

**Comment by GeorgeWeb** [comment]
> @GeorgeWeb Could you please address CI failure on Windows? FAIL: Clang :: Driver/sycl-cuda-rdc.cpp (10320 of 18711) ******************** TEST 'Clang :: Driver/sycl-cuda-rdc.cpp' FAILED ******************** Exit Code: 1

@againull 
I don't have a way to test Windows other than trying here on the CI, hence I opted for marking the test unsupported for Windows. Not seeing anything that could trigger the issue, I am wondering if you've an idea whether I need to add something Windows specific to the `clangxx` command in a Windows-specific version of the lit test to have it passing on Windows too? If not, I'll have it skip running on Windows for now.

**Comment by againull** [comment]
> > @GeorgeWeb Could you please address CI failure on Windows? FAIL: Clang :: Driver/sycl-cuda-rdc.cpp (10320 of 18711) ******************** TEST 'Clang :: Driver/sycl-cuda-rdc.cpp' FAILED ******************** Exit Code: 1
> 
> @againull I don't have a way to test Windows other than trying here on the CI, hence I opted for marking the test unsupported for Windows. Not seeing anything that could trigger the issue, I am wondering if you've an idea whether I need to add something Windows specific to the `clangxx` command in a Windows-specific version of the lit test to have it passing on Windows too? If not, I'll have it skip running on Windows for now.

Disabling the test seems right as it is cuda specific, we don't have windows systems with cuda installation in CI.

**Pull request merged**